### ISSUE TITLE
Ideas for V3.0

### ISSUE THREAD

**Issue by Fireboyd78** [crosire/scripthookvdotnet#449]
Since we're already at V2.7.x, it's probably a good idea to start planning for V3.0. This will be a major version, so we should throw backwards compatibility straight out of the window and take the opportunity to remove all obsolete methods/properties and rename ones that might have a strange name or not accurately reflect their purpose.

More ideas will be added periodically, so feel free to leave your own idea/feedback below.
# Native hashes

One of the biggest changes will definitely be the native hashes. There's a lot of duplicates or ones with bad names from months and months ago. Since there's literally thousands of hashes, I suggest splitting NativeHashes.hpp into separate files like so:

`[core]\hashes\[category].hpp` -- partial `Native::Hash` enum definitions
`[core]\NativeHashes.hpp` -- includes all partial definitions from `NativeHashes` folder

There will be multiple files containing native hashes, but they will still be all in one enum.
# Attributes
## ScriptHookVersionAttribute

Also new for 3.0 should be the ability for developers to implement a `ScriptHookVersion` attribute that allows user to define a Minimum and a Maximum supported version of ScriptHookV.NET. There's actually a built-in class called `System.Version` that can be used to process these values. The attribute constructor should support all of the `System.Version` constructors. The `Revision` part of the Version should be ignored.

This attribute should only be applicable to the assembly of the script, and accessed when the script is first loaded. Any scripts that are not safely compatible should be unloaded and a message printed to the log file.
# Classes
- [x] Implement a **Checkpoint** class [#486]
- [x] Implement a **ParticleFX** (_PTFX_) class [#535]
# Source code
## XML Documentation

We should start getting into the habit of adding XML documentation to methods/properties/enums so developers don't have to try and guess what something does.

**Comment by crosire** [comment]
I like the version attribute idea a lot. We already have a ```RequireScript``` attribute for scripts to manage dependencies, a ```RequireScriptHookVersion``` attribute or similar would be great here. This could even go as far as the library downloading and referencing an older version for scripts that weren't updated yet.

**Comment by kagikn** [comment]
[Here's PluginAttribute in RAGE Plugin Hook](http://docs.ragepluginhook.net/html/FCA1A76.htm). I think we can bring some features from it, such as the author property.

**Comment by JohnnyCrazy** [comment]
I also like the idea of a Version Attribute, 

> This could even go as far as the library downloading and referencing an older version for scripts that weren't updated yet.

So we would load more versions of scripthookdotnet at once?

**Comment by kagikn** [comment]
I also think the idea of a VersionAttribute is very good idea.

**Comment by Fireboyd78** [comment]
@kagikn I think you missed the point about not having any backwards compatibility from V3.0 and onwards. All 3.x scripts should require a version attribute with a minimum version, or else it will not be loaded. This helps prevent people reporting "bugs" while trying to load 2.x scripts with 3.x ScriptHook. If mod developers want their mods to stay relevant, it should be their responsibility to make sure it's 3.x compatible.

Nothing is set in stone yet obviously, but I really hate supporting backwards compatibility. It clutters up code and can lead to unintentional bugs.

**Comment by kagikn** [comment]
@Fireboyd78 Yeah, I missed that point, but I'm really worried people complain about it. (if we decide not to support any backwards compatibility on 3.x)

**Comment by crosire** [comment]
The point of major updates is that they are allowed to break backwards compatibility. There are a few points I'd like to change in the code base, which I can't right now, because they would break existing scripts.
But we also cannot just push out a new update which basically makes all existing scripts incompatible without providing a solution for those that won't be updated anymore. There needs to be a temporary solution. One option would be to push an update to the 2.x branch which checks for a "RequireScriptHookVersion" and only executes script where it does not exist. This way both 2.x and 3.x can be installed at the same time and each will only executive the scripts written against it.

Another idea would be to set the .NET assembly version and check that one when loading a script assembly and resolving assemblies. We'll probably get a bunch of errors when loading old scripts otherwise, because they reference methods and properties that no longer exist.

**Comment by Fireboyd78** [comment]
We could split the native hashes into separate files, then put all of the legacy ones into the main `NativeHashes.hpp` file and add an `ObsoleteAttribute` to them so developers can update their code.

I'm also willing to write some documentation for developers that tells them exactly how to convert their 2.x scripts to 3.x, which would pretty much be a list of stuff that's been removed and what should be used in its place, as well as a list of methods/properties that have a different behavior (if any are applicable).

> Another idea would be to set the .NET assembly version of this project and check that one when loading a script assembly and resolving assemblies, rather than settings a custom attribute.

I'm not quite sure what you mean by that?

**Comment by crosire** [comment]
```
using namespace System::Reflection;

[assembly:AssemblyTitleAttribute("Community Script Hook V .NET")];
[assembly:AssemblyCopyrightAttribute("Copyright © 2016")];
[assembly:AssemblyVersionAttribute("3.0.0")];
```

**Comment by crosire** [comment]
I don't think we should split the native hashes into separate files by the way. Yes, it's a big file, but some clean up should be enough (like removing all the "_0x..." enum values).

**Comment by crosire** [comment]
And a big no to the "Tabs2Spaces". I'm a huge fan of tabs :wink:.

**Comment by kagikn** [comment]
I very agree with adding XML documentation, RPH has better one. I don't agree with "Tabs2Spaces", though :neutral_face:.

**Comment by Fireboyd78** [comment]
I think there comes a point where there's simply too many native hashes in one file...there's close to 9,000 lines of code in there, and that's not including any XML documentation that may be added.

I honestly cannot think of any reason why you guys would want to keep all the native hashes in one huge file that slows down Visual Studio and even BitBucket's viewer...

**Comment by Fireboyd78** [comment]
@kagikn I think it's better for RPH to be its own thing instead of ripping ideas from them...I mean seriously, I've realized lately a lot of the stuff you commit to this project are just rips from RPH (IHandleable/ISpatial are two examples). Why not be original?

**Comment by crosire** [comment]
I agree. I didn't even know RPH had ```IHandleable``` and ```ISpatial``` This is a separate project that existed before and should therefore continue to innovate on its own.

@Fireboyd78: That file is (at least it used to be) generated automatically with a tool by @JohnnyCrazy using http://dev-c.com/nativedb/downloads.

**Comment by kagikn** [comment]
I dont think it's good idea to rip everything from RPH without considering values, we should innovate on its own, yes.

**Comment by kagikn** [comment]
How about `ScriptFire` class, though .Net Script Hook for GTA IV has this feature, RPH doesn't.

**Comment by Fireboyd78** [comment]
I stand corrected on the console, my bad. And well, if the tool can automagically generate hashes while also adding "obsolete" attributes to old hash names, then it might work.

**Comment by JohnnyCrazy** [comment]
Should be possible, currently the tool lives [here](https://github.com/JohnnyCrazy/scripthookvdotnet/tree/native-generator), but I guess we could trasfer it to another repo.

**Comment by Fireboyd78** [comment]
I'd also like to suggest removing all properties that only have a setter and instead replacing them with methods. If the getter is ever found, we could always deprecate the setter method.

A property cannot be a property if you can only assign to it, lol.

**Comment by Daniel15** [comment]
> We should start getting into the habit of adding XML documentation to methods/properties/enums so developers don't have to try and guess what something does.

Once you're at a point where all public methods are documented (even with stubs inserted via something like GhostDoc), there's lint rules you can use to ensure that the public API remains fully documented. I'm pretty sure StyleCop has a rule for it.

**Comment by Daniel15** [comment]
> We should start getting into the habit of adding XML documentation to methods/properties/enums so developers don't have to try and guess what something does.

Once you're at a point where all public methods are documented (even with stubs inserted via something like GhostDoc), there's lint rules you can use to ensure that the public API remains fully documented. I'm pretty sure StyleCop has a rule for it.

> A property cannot be a property if you can only assign to it, lol.

Set-only properties are still useful for object initializer syntax.

**Comment by zorg93** [comment]
I for one think a console would be needed, that way you could do unload and reload specific scripts on the fly. As well as print script errors to the user so they know if a script actually stopped working without having to alt tab out and check the log file

**Comment by JohnnyCrazy** [comment]
We ditched the idea before because a smiliar thing exists [here](https://github.com/Dakota628/GTAVDeveloperConsole)

**However**, I aggree that an integrated console would be useful. Especially if you quickly want to check some log-output or test a specific thing via a command (instead of always having to create a hotkey). Also, reloading specific scripts sounds very useful to me, especially if you have larger ones.

I also thought about it being a seperate (script-)project, but this will lead to problems:
* We shouldn't allow disabling/enabling scripts from other scripts.
* Errors could be read via `FileSystemWatcher`, but that sounds hacky.

What's your current opinion on this? @crosire

**Comment by zorg93** [comment]
An external console just wouldn't work. Unloading and reloading scripts can only be done by the hook itself. If a console is made should scripts be able to send commands to the console or just the user. If scripts have access to it then there is potential for scripts terminating other scripts. 
I'm also thinking you could incorporate a variable viewer.
Like print ExampleScript.ExampleVar or use reflection to get all vars.  Though it would only work for public variables

**Comment by zorg93** [comment]
Could we also add an OnTerminated eventhandler for scripts that would be invoked when the insert key gets pressed. That way you could clean up (deleting entites/blips, resetting cameras, resetting player model, saving script settings etc) before the scriptdomain gets killed

**Comment by crosire** [comment]
@zorg93 Isn't the destructor already called then?

**Comment by zorg93** [comment]
Nope, well if it is, it certainly doesnt let natives be called

**Comment by crosire** [comment]
But it would be the preferred way to solve this I think. Will need to check if that can be made working.

**Comment by zorg93** [comment]
I think what happens is the script domain gets killed before the scripts have been disposed so they can't actually run the disposal

**Comment by zorg93** [comment]
I tried manually calling the script destructor in the Script::Abort method, but a class that inherits scripts wont get its destructor called

**Comment by zorg93** [comment]
Im pretty sure an event is the best way to go about this.

**Comment by crosire** [comment]
It's the easiest way. We could also inherit from ```IDisposable```. But an event is probably the most obvious. Could be called something like "Aborted".

**Comment by zorg93** [comment]
Hmm, just found out even though the scripts dont implicitly inherit IDisposable, you can still use `		protected override void Dispose(bool Disposing)
		{		 
			if (Disposing)
			{
				UI.ShowSubtitle("Testing");
			}
		}`
I may aswell roll back the Aborted Event

**Comment by JohnnyCrazy** [comment]
IMO we should stick to the event, easier for people and since intelligence doesn't even recognize the dispose override, it's more obvious

**Comment by crosire** [comment]
I agree.  It's debatable whether ```Dispose``` would be the proper place for clean up anyway (it's mainly designed for destroying unmanaged resources, no complicated logic).

**Comment by zorg93** [comment]
Well then I guess Aborted is here for good

**Comment by crosire** [comment]
Removed ```IDisposable``` in 47779f7feb08196aa495493f7d2820da6b357644 to prevent any further misunderstandings.

**Comment by JohnnyCrazy** [comment]
@crosire so, what about that console idea :P

**Comment by crosire** [comment]
@JohnnyCrazy If it's simple and small, than yeah, why not. If somebody is willing to implement it.

I suggest we rename the new ASI to ```ScriptHookVDotNet3.asi``` (or similar) and add an assembly version attribute. This way we can check against the version when resolving and it's possible to install version 2 and version 3 alongside each other (so old scripts can still be used).

**Comment by zorg93** [comment]
Is it really possible to install 2 different versions of SHV.Net, would've thought the clr would have some issues about that. Also the idea of ScriptHookVDotNet3 doesnt sound that great imo, it sounds like a temporary solution that would cause further problems down the line

**Comment by crosire** [comment]
Shouldn't be a problem. The CLR is loaded only once.
We need to provide some temporary solution. This would be one.

**Comment by zorg93** [comment]
I just cant think of a nice way to set up 2 working with each other. Also dlls, like nativeUI, would probably not be working across the 2 different assemblies.

**Comment by Fireboyd78** [comment]
How about V3 continues to use `ScriptHookVDotNet` and V2 will use `ScriptHookVDotNet_Legacy`?

**Comment by zorg93** [comment]
Is anyone planning on implementing the console, I am thinking about asking Dakota628(the guy who made the developer console script) if we could copy parts of that into the new version of script hook to get a working framework to build on top off

**Comment by JohnnyCrazy** [comment]
@zorg93 maybe we should build a little concept first? So we can confirm what we need and what not

**Comment by Fireboyd78** [comment]
@crosire @JohnnyCrazy seems like we have plenty of good ideas for v3 -- is this a good time to create a "v3_dev" branch so we can start pushing commits?

**Comment by crosire** [comment]
I guess so, yeah.

**Comment by crosire** [comment]
2.X and 3.X development is now moved into two separate branches, ```dev_v2``` and ```dev_v3```. The default ```dev``` branch currently points to ```dev_v2```, but will be updated to ```dev_v3``` once it's ready.

**Comment by zorg93** [comment]
Checkpoint class was added in #486. Particle Effects have been added(subject to discussion) in #535

**Comment by Fireboyd78** [comment]
Thank you @zorg93, I've updated the OP.

**Comment by sollaholla** [comment]
- Start scenario in place method on peds `ped.Task.StartScenarioInPlace(string scenario);`
- Start scenario with heading as an argument. (Please..)
- Warp to nearest scenario? `ped.Task.WarpToNearestScenario(Vector3 position);`
- Setting props to dynamic: `prop.Dynamic = true;`
- Disabling radar: `Game.DisableRadarThisFrame();`
- It'd be nice if you can add events to tasks `ped.OnPerformedTask += MyFunction;`, not sure how it would look.

**Comment by RainerHilmer** [comment]
I don't know if this has been discussed already, so here I go anyway: Most .net script mods are not maintained anymore. Nevertheless there are some very great and useful mods out there and it would be sad to see them not working anymore because of the breaking changes which shvdn3 introduces. So my idea is to give shvdn3 a different assembly name. This way both (2.9.2 and 3) could co-exist. Mod developers who are still active on their mods could upload a second version, one for 2.9.2 and another for 3.

**Comment by crosire** [comment]
V2 and V3 already can coexist. V2.9.2 doesn't load 3.0 scripts and V3.0 doesn't load V2.9.2 scripts. This was done for exactly the reason you mentioned.

**Comment by RainerHilmer** [comment]
Okay but
1. how can both co-exist if they have the same filename?
2. how does shvdn differentiate between a script for 2.9.2 and 3 other than by throwing an exception? What does a mod developer need to do? Is there an Attribute we have to use or what's the deal? I am somewhat clueless at this point.

**Comment by Fireboyd78** [comment]
@RainerHilmer The old version will have a different filename, e.g. "ScriptHookVDotNet_V2". It differentiates between scripts by [checking the referenced assembly version of SHVDN](https://github.com/crosire/scripthookvdotnet/blob/dev_v2/source/core/ScriptDomain.cpp#L75). [V3 will try loading any version scripts](https://github.com/crosire/scripthookvdotnet/blob/dev_v3/source/core/ScriptDomain.cpp#L86) (not sure why though).

**Comment by RainerHilmer** [comment]
Changing the name of the old shvdn is a bad idea imo because it would break old script mods. They would simply don't find the assembly anymore. I think it would be better to name V3 ScriptHookVDotNet3 or maybe even shvdn3. The abbreviation shvdn has become pretty common in the developer community.

**Comment by kagikn** [comment]
Changing the name of shvdn isn't cause a serious problem, changing the name of the script folder for shvdn is much bigger problem. Because many scripts don't care whether the script folder name is changed from "scripts".

**Comment by crosire** [comment]
@RainerHilmer No it's not bad. It does make zero difference actually. You can name the assembly whatever you want due to the way the script loading of SHVDN works.

**Comment by RainerHilmer** [comment]
I don't talk about the script loading mechanism of shvdn but about the framework references. If a script written for 2.9.2 references ScriptHookVDotNet.dll it will not work anymore if 2.9.2 gets renamed to ScriptHookVDotNet_old.dll and V3 gets the name ScriptHookVDotNet.dll instead. This script still references ScriptHookVDotNet.dll and therefore calls into the code of V3.

**Comment by zorg93** [comment]
That doesn't matter, the dll is only used for devs when writing a script.

**Comment by crosire** [comment]
SHVDN bypasses the normal assembly referencing system, so it's no issue.

**Comment by RainerHilmer** [comment]
@zorg93: I am one of those devs and I do write scripts. That's why I am here. :)
@crosire: okay shvdn does this but not those old script mods.

**Comment by crosire** [comment]
You don't understand. SHVDN bypasses it **for** the script assemblies. Just believe me. It works. I use it myself (so to have both scripts for 2.9.2 and do testing with 3.0).

**Comment by RainerHilmer** [comment]
So what happens if (in the future) someone installs shvdn 3, therefore overwrites shvdn 2.9.2 because of the same filename and tries to run a script mod which was written for 2.9.2?

**Comment by crosire** [comment]
SHVDN will be distributed as `ScriptHookVDotNet3.dll` or something similar.

**Comment by Fireboyd78** [comment]
All in favor of `ScriptHookVDotN3T.dll`, say LOL! :P

**Comment by RainerHilmer** [comment]
Oh and I hope there will be no namespace conflict with shvdn 2. You know we have to keep both versions or older mods which aren't updated anymore will cease to run and that would be a bummer because there are some real good ones out there. Old but great.

**Comment by crosire** [comment]
As explained before the two can be used together without any conflicts.

**Comment by Nacorpio** [comment]
What else do we need to  add before we can release the SHDN v3?

**Comment by crosire** [comment]
The console has still some quirks and I experienced some crashes when starting with SHVDN v3. I don't have the time to look into these right now though, so it's on hold.

**Comment by RainerHilmer** [comment]
Apparently we can not use them together. I forked this repo and compiled it. The logs tell me that ScriptHookVDotNet3 and ScriptHookVDotNet are properly hooked but when I insert a script written for 3 into the scripts folder (among scripts written for 2) the game hangs in the loading process. I doesn't crash but it just loads infinitely.

**Comment by crosire** [comment]
Version v2 and v3 should work side-by-side again now, so I'll probably release v3 in the next weeks.

**Comment by zorg93** [comment]
Only a year in :P

**Comment by Nacorpio** [comment]
It's such a shame seeing this project die.

**Comment by JohnnyCrazy** [comment]
@Nacorpio Why would you consider it in a dieing state? It's currently stable, used by a lot of SP and MP mods (I mean literally custom MP mods) and crosire/all try to keep the issue-count low.

**Comment by crosire** [comment]
Correction on my previous post: Multiple SHVDN versions sadly still don't work side-by-side, because multiple fibers hosting the CLR confuse the system and cause a stack corruption on startup. I'm a bit lost on how to solve that in the moment. And without that working v3 cannot be released.

**Comment by RainerHilmer** [comment]
@crosire Just to avoid confusion why I gave your comment a thumbs up: I welcome your decision. :)

**Comment by Fireboyd78** [comment]
@crosire Would it be possible to host one instance of the CLR and have it handle both SHVDN versions at once? Or even have a new, separate DLL that serves as the .ASI and then that ASI will load the multiple SHVDN versions using a single instance of the CLR?

**Comment by crosire** [comment]
The separate CLR hosting idea could potentially work. The whole thing is a bit complicated due to both versions containing symbols that would very likely clash if loaded into the same appdomain.

**Comment by RainerHilmer** [comment]
A radical idea: How about you guys implement all of V3 into V2 (and forget about V3) even if it means partly redundant code? I know this is usually a no-go for developers craving for clean code ( I am a clean code developer myself), but if there's no other way...
You have already the Abort event of V3 in V2 and still maintain the Dispose method in V2. Why not carry on this way? Just mark all outdated things with the Obsolete attribute (w/o throwing an Exception).

**Comment by crosire** [comment]
IDK. I feel like V3 will never be released. Even though it has so many awesome features. But no backwards compatibility. And implementing those features into V2 isn't that easy either, because the backend is so different.

**Comment by JohnnyCrazy** [comment]
@crosire No release would be 😢 

I think it's better to release it at some point. If some developer doesn't update their mod, its users will have to live with it, shit happens I guess.

**Comment by Fireboyd78** [comment]
I agree with @JohnnyCrazy. Set a deadline for developers to update their mods, and after the deadline passes, release a hotfix for V2 that maybe adds a popup box like this:

> Thank you for your continued support of ScriptHookV.NET! This one-time message is to inform you that V2 of our mod is no longer supported, and we highly recommend you update to the latest version to take full advantage of the latest features!
Don't worry, this message won't appear ever again. Just press 'OK' to continue normally. Thanks again! :)

**Comment by RainerHilmer** [comment]
Most of the great script mods have been abandoned, unfortunately. It would be fatal if they don't run anymore because of breaking changes! I'd rather NOT update to V3 if that happens. So this is not about making it right for everyone but about disappointing most gamers.

**Comment by crosire** [comment]
I agree. Abandoning support for V2 mods would be a disaster. I still hope a good solution is found which allows proper side-by-side usage of both versions. I really want to release V3. I very much prefer its codebase over V2. But my hands are tied until me or somebody else finds the time and manages to fix compatibility with the old version.

**Comment by crosire** [comment]
Just pushed commit 1f140898f6752096a7e54b22e3dd5afc69e5e50f which might solve the problem by providing a single CLR entry point / ASI that can load and invoke multiple SHVDN versions. Couldn't test it yet though.

**Comment by Nacorpio** [comment]
Will the recent illegalization of mods affect the development process of SHDN?

**Comment by zorg93** [comment]
The main issue is whether or not SHV will become updated or not

**Comment by Nacorpio** [comment]
It probably will not, so I guess we are doomed.

**Comment by zorg93** [comment]
probably. If it does die a replacement will most likely not support earlier game versions

**Comment by JohnnyCrazy** [comment]
We'll have to wait for a statement of Alexander, everything else is pure speculation. But TBH since T2 wants to get rid of OpenIV, I'm pretty sure they don't want scripthook around any more too.

**Comment by Nacorpio** [comment]
It's funny because Open IV isn't the problem.
It was singleplayer-only.

**Comment by zorg93** [comment]
sure people could adapt open hook for 1103 though

**Comment by kagikn** [comment]
@crosire Should we close this issue? Checkpoint doesn't seem to work for me though (but needs more tests to conclude), but I have no idea how to fix the class if the memory pattern is wrong.

**Comment by crosire** [comment]
I'd keep this issue open until v3 is released. It's more of a general discussion thread at this point.

**Comment by kagikn** [comment]
Okay, that makes sense.

**Comment by varunkumar2310** [comment]
Hi, we need to have categories for vehicles and ped (like in menyoo) (if possible other properties)
Once i tried to spawn a random vehicle and a ped in it on a normal road. The game spawned a helicopter with a whale attached to its driver seat. :P
In menyoo they have "PedList.xml"  and "VehicleList.xml".
We can use that as a reference to create categories for these.
